backend/app/tools.py

```diff
@@ -287,7 +289,16 @@ def _get_tavily_answer():
 
 
 def _get_action_server(**kwargs: ActionServerConfig):
-    toolkit = ActionServerToolkit(url=kwargs["url"], api_key=kwargs["api_key"])
+    def _build_context(secrets: dict) -> str:
+        ctx = json.dumps({"secrets": secrets})
+        return base64.b64encode(ctx.encode("utf-8")).decode("ascii")
```

Comment:
What would be actually stored in the database? It would be maybe easier to assume that secrets is already in the right format and do the base64 encoding etc before storing the data.

Reply:
This PR takes the approach of config looking this way:

```python
class ActionServerConfig(ToolConfig):
    url: str
    api_key: str
    secrets: Optional[dict]
```

I followed [these docs](https://github.com/Sema4AI/actions/blob/readme-update/action_server/docs/guides/07-secrets.md) for context encoding.

Here's an example assistant config in the database:

```
{
    "configurable": {
        "type": "agent",
        "type==agent/agent_type": "GPT 3.5 Turbo",
        "type==agent/interrupt_before_action": false,
        "type==agent/retrieval_description": "Can be used to look up information that was uploaded to this assistant.\nIf the user is referencing particular files, that is often a good hint that information may be here.\nIf the user asks a vague question, they are likely meaning to look up info from this retriever, and you should call it!",
        "type==agent/system_message": "You are a helpful assistant.",
        "type==agent/tools": [
            {
                "id": "f7015ef0-5a84-4da0-b2f1-f37ba09c548d",
                "type": "action_server_by_sema4ai",
                "name": "Action Server by Sema4.ai",
                "description": "Run AI actions with [Sema4.ai Action Server](https://github.com/Sema4AI/actions).",
                "config": {
                    "url": "https://sixty-five-chatty-rats.robocorp.link",
                    "api_key": "cbHz3L1yx8_0SMBsuqjcEtDvY4CZEm2HuJSMmT-U8oY",
                    "secrets": {
                        "mypwd": "green"
                    }
                }
            }
        ],
        "type==chat_retrieval/llm_type": "GPT 3.5 Turbo",
        "type==chat_retrieval/system_message": "You are a helpful assistant.",
        "type==chatbot/llm_type": "GPT 3.5 Turbo",
        "type==chatbot/system_message": "You are a helpful assistant."
    }
}
```